brainfuck/bf.zig

```diff
@@ -0,0 +1,189 @@
+        var self = Tape{ .tape = std.ArrayList(i32).init(alloc) };
+        self.tape.append(0) catch unreachable;
+        return self;
+    }
+
+    fn get(self: *const Tape) i32 {
+        return self.tape.items[self.pos];
+    }
+
+    fn inc(self: *Tape, x: i32) void {
+        self.tape.items[self.pos] += x;
+    }
+
+    fn move(self: *Tape, x: i32) void {
+        self.pos += @intCast(usize, x);
+        if (self.pos >= self.tape.items.len) {
+            self.tape.appendNTimes(0, self.tape.items.len * 2) catch unreachable;
+        }
+    }
+};
+
+const Ops = std.ArrayList(Op);
+
+const Op = struct {
+    op: OpType = undefined,
+    val: i32 = 0,
+    loop: Ops = undefined,
+};
+
+const StrIterator = struct {
+    text: []const u8,
+    pos: usize = 0,
+
+    fn next(self: *StrIterator) ?u8 {
+        if (self.pos < self.text.len) {
+            const res = self.text[self.pos];
+            self.pos += 1;
+            return res;
+        }
+        return null;
+    }
+};
+
+const Program = struct {
+    ops: Ops,
+    p: *Printer,
+    a: *std.mem.Allocator,
+
+    fn init(alloc: *std.mem.Allocator, code: []const u8, printer: *Printer) Program {
+        return Program{
+            .ops = parse(alloc, &StrIterator{ .text = code }),
+            .p = printer,
+            .a = alloc,
+        };
+    }
+
+    fn run(self: *const Program) void {
+        var tape = Tape.init(self.a);
+        self._run(&self.ops, &tape);
+    }
+
+    fn parse(alloc: *std.mem.Allocator, iter: *StrIterator) Ops {
+        var res: Ops = Ops.init(alloc);
+        while (iter.next()) |value| {
+            switch (value) {
+                '+' => res.append(Op{.op = OpType.INC, .val = 1,}) catch unreachable,
+                '-' => res.append(Op{.op = OpType.INC, .val = -1,}) catch unreachable,
+                '>' => res.append(Op{.op = OpType.MOVE, .val = 1,}) catch unreachable,
+                '<' => res.append(Op{.op = OpType.MOVE, .val = -1,}) catch unreachable,
+                '.' => res.append(Op{.op = OpType.PRINT,}) catch unreachable,
+                '[' => res.append(Op{.op = OpType.LOOP, .loop = parse(alloc, iter)}) catch unreachable,
+                ']' => return res,
+                else => continue,
+            }
+        }
+        return res;
+    }
+
+    fn _run(self: *const Program, program: *const Ops, tape: *Tape) void {
+        for (program.items) |*op| {
+            switch (op.op) {
+                OpType.INC => tape.inc(op.val),
+                OpType.MOVE => tape.move(op.val),
+                OpType.LOOP => while (tape.get() > 0) self._run(&op.loop, tape),
+                OpType.PRINT => self.p.print(tape.get()),
+            }
+        }
+    }
+};
+
+fn notify(msg: []const u8) void {
+    const addr = std.net.Address.parseIp("127.0.0.1", 9001) catch unreachable;
+    var stream = std.net.tcpConnectToAddress(addr) catch unreachable;
+    _ = stream.write(msg) catch unreachable;
+    stream.close();
```

Comment:
I could miss it before, but what would happen if no server is available (like if the binary is executed as-is)? Also, maybe `defer` could be used here.

Reply:
That's actually a good point, it should probably crash as it is, as there's an error actually being returned when there's no server. Surprisingly it does not... So what should be the behavior here? Should it crash, print an error and exit, ignore error and run test?

Reply:
Other tests ignore any networking issues, so "ignore error and run test" would be consistent with other code.

Reply:
I left it as is, because it works as expected with current build settings. If we were to use e.g. ReleaseSafe, then the program would panic. I suggest we keep it as is, and I'll create a separate patch to handle possible error in a nicer way (for all zig tests).